### PR TITLE
8295788: C2 compilation hits "assert((mode == ControlAroundStripMined && use == sfpt) || !use->is_reachable_from_root()) failed: missed a node"

### PR DESCRIPTION
**Pull request by rwestrel**
This failure is similar to previous failures with loop strip mining: a
node is encountered that has control set in the outer strip mined loop
but is not reachable from the safepoint. There's already logic in loop
cloning to find those and fix their control to be outside the
loop. Usually a node ends up in the outer loop because some of its
inputs is in the outer loop. The current logic to catch nodes that are
erroneously assigned control in the outer loop is to start from
safepoint's inputs and look for uses with incorrect control. That
doesn't work in this case because: 1) the node is created by
IdealLoopTree::reassociate in the outer loop because its inputs are
indeed there 2) but a pass of split if updates the control to be
inside the inner loop.

To fix this, I propose reusing the existing clone_outer_loop_helper()
but apply it to the loop body as well. I had to tweak that method
because I ran into cases of dead nodes still reachable from a node in
the loop body but removed from the _body list by
IdealLoopTree::DCE_loop_body() (and as a result not cloned).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295788](https://bugs.openjdk.org/browse/JDK-8295788): C2 compilation hits "assert((mode == ControlAroundStripMined && use == sfpt) || !use->is_reachable_from_root()) failed: missed a node"

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [32271652](https://git.openjdk.org/jdk/pull/11162/files/322716529b18b5fbfcac34956cdb5bb633a2df3d)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11162/head:pull/11162` \
`$ git checkout pull/11162`

Update a local copy of the PR: \
`$ git checkout pull/11162` \
`$ git pull https://git.openjdk.org/jdk pull/11162/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11162`

View PR using the GUI difftool: \
`$ git pr show -t 11162`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11162.diff">https://git.openjdk.org/jdk/pull/11162.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Make sense.

File: test/hotspot/jtreg/compiler/loopstripmining/TestUseFromInnerInOuterUnusedBySfpt.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            i11 = 14;
+            do {
+                i5 |= i6;
+            } while (--i11 > 0);
+        }
+    }
+
+    public static void main(String[] strArr) {
+        TestUseFromInnerInOuterUnusedBySfpt _instance = new TestUseFromInnerInOuterUnusedBySfpt();
+        for (int i = 0; i < 10; i++ ) {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopstripmining/TestUseFromInnerInOuterUnusedBySfpt.java:57]:
```suggestion
        for (int i = 0; i < 10; i++) {
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 2]
```diff
@@ -2297,6 +2299,11 @@ void PhaseIdealLoop::clone_outer_loop(LoopNode* head, CloneLoopMode mode, IdealL
       clone_outer_loop_helper(old, loop, outer_loop, old_new, wq, this, true);
     }
 
+    for (uint i = 0; i < loop->_body.size(); i++) {
+      Node* old = loop->_body.at(i);
+      clone_outer_loop_helper(old, loop, outer_loop, old_new, wq, this, true);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:2304]: While you're at it, could you rename the helper method to something more meaningful?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

File: src/hotspot/share/opto/loopopts.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopopts.cpp]: Thanks for reviewing this. I did in the new commit. Does it look good to you?

**Review by TobiHartmann — APPROVED** [review]
Looks good, thanks for updating!

**Comment by rwestrel** [comment]
@vnkozlov @TobiHartmann thanks for the reviews.

**Pull request closed without merging**